### PR TITLE
fix: fix migration 105 to recover from undefined selected address

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**

This pull request fixes the selectedAddress migration issue where the selectedAddress is not a string when migrating to internal accounts. In this broken state, the first identity would then be used for the `selectedInternalAccount` and `selectedAddress`

## **Related issues**

Fixes: [383](https://github.com/MetaMask/accounts-planning/issues/383)

## **Manual testing steps**

1 . Comment out all the migrations from 105 and onwards.
2. Reinstall the extension and create 2 accounts, and select the second account. Dump the metamask state and see that the second account is the selectedAddress
3. In migration 105, force the selectedAddress to be undefined.
4. Uncomment the migrations
5. Restart the extension, see that the selectedAddress  is now the first account

### **For release testing**

1. Install previous version of MM extension
2. Create accounts (import private or connect hardware wallet)
3. Edit the names of some accounts
4. Update MM extension to the release version
5. The state of the accounts should be the same as before

## **Screenshots/Recordings**

Not applicable

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/migrations/105.test.ts [hunk 1]
```diff
@@ -285,5 +296,41 @@ describe('migration #105', () => {
...
+      expect(newStorage.data).toStrictEqual({
+        PreferencesController: expect.objectContaining({
+          selectedAddress: MOCK_ADDRESS,
+        }),
+        AccountsController: {
+          internalAccounts: {
+            accounts: {
+              [expectedUUID]: expectedInternalAccount(
+                MOCK_ADDRESS,
+                `Account 1`,
```
Comment (ccharly) [inline app/scripts/migrations/105.test.ts:327]: I would have expect the linter to complain about this one (backtick vs `'`) while there's no "variables to expand"

File: app/scripts/migrations/105.test.ts [hunk 2]
```diff
@@ -285,5 +296,41 @@ describe('migration #105', () => {
...
+      const oldData = {
+        PreferencesController: {
+          identities: {
+            [MOCK_ADDRESS]: { name: 'Account 1', address: MOCK_ADDRESS },
+          },
+          selectedAddress: undefined,
+        },
+      };
+      const oldStorage = {
+        meta: { version: 103 },
```
Comment (ccharly) [inline app/scripts/migrations/105.test.ts:312]: 103 -> 104

File: app/scripts/migrations/105.ts [hunk 3]
```diff
@@ -106,26 +127,33 @@ function createSelectedAccountForAccountsController(
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   state: Record<string, any>,
 ) {
-  const selectedAddress = state.PreferencesController?.selectedAddress;
+  const selectedAddress = undefined; // state.PreferencesController?.selectedAddress;
```
Comment (ccharly) [inline app/scripts/migrations/105.ts:130]: Shouldn't we keep the `PreferencesController.selectedAddress` here? If users do not face the issue (of having no selected address), then the migration would go through the "naturally".

We only want to cover users that fall into the "undefined" case, no?

**Review by ccharly — COMMENTED** [review]
Small proposal, what about re-writing this function like:

```ts
function getFirstAddress(
  // TODO: Replace `any` with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  state: Record<string, any>,
) {
  const [firstAddress] = Object.keys(
    state.PreferencesController?.identities || {},
  );
  return firstAddress;
}

function createSelectedAccountForAccountsController(
  // TODO: Replace `any` with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  state: Record<string, any>,
) {
  let selectedAddress = state.PreferencesController?.selectedAddress;
  
  if (typeof selectedAddress !== 'string') {
    global.sentry?.captureException?.(
      new Error(
        `state.PreferencesController?.selectedAddress is ${selectedAddress}`,
      ),
    );

    // Get the first account if selectedAddress is not a string
    selectedAddress = getFirstAddress(state);
  }

  const selectedAccount = findInternalAccountByAddress(state, selectedAddress);
  if (selectedAccount) {
    // Required in case there was no address selected
    state.PreferencesController.selectedAddress = selectedAccount.address;
    state.AccountsController.internalAccounts = {
      ...state.AccountsController.internalAccounts,
      selectedAccount: selectedAccount.id,
    };
  }
}
```

**Review by ccharly — APPROVED** [review]
LGTM!

**Review by danroc — APPROVED** [review]

**Review by danroc — APPROVED** [review]

**Review by danroc — APPROVED** [review]
Tested in a call with @montelaidev.

**Pull request merged**